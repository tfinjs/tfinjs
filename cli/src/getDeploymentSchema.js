import fromPairs from 'lodash/fromPairs';
import getBuiltResources, {
  STATUS_DESTROYED,
  STATUS_OUTDATED,
  STATUS_UNDEPLOYED,
} from './getBuiltResources';

const getDeploymentSchema = (project, fs) => {
  /*
    add: everything that should be desployed (no deployment_status.txt),
    update: with deployment_status.txt
    remove: everything that has deployment_status.txt and deployment_status.txt !== DESTROYED
  */

  const dist = project.getDist();
  
  const graph = project.getDependencyGraph();

  const resources = fromPairs(
    project.getResources().map((resource) => [
      resource.versionedName(),
      {
        uri: resource.getUri(),
        contentHash: resource.getContentHash(),
      },
    ]),
  );
  const currentResourceKeys = Object.keys(resources);

  const builtResources = Object.entries(getBuiltResources(dist, fs));

  const m = ([a]) => a;

  const add = builtResources
    .filter(
      ([name, { status }]) =>
        [STATUS_UNDEPLOYED].includes(status)
        && currentResourceKeys.includes(name),
    )
    .map(m);

  const update = builtResources
    .filter(
      ([name, { status }]) =>
        [STATUS_OUTDATED].includes(status) && currentResourceKeys.includes(name),
    )
    .map(m);

  const remove = builtResources
    .filter(
      ([name, { status }]) =>
        ![STATUS_DESTROYED, STATUS_UNDEPLOYED].includes(status)
        && !currentResourceKeys.includes(name),
    )
    .map(m);


  return {
    resources,
    add,
    remove,
    update,
    graph,
  };
};

export default getDeploymentSchema;
