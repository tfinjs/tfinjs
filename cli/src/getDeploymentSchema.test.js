/* eslint-env jest */
import makeProject from '../mocks/sampleProject';
import getDeploymentSchema from './getDeploymentSchema';

test('getDeploymentSchema', () => {
  const {
    project, vol, fs, table,
  } = makeProject();

  const schema = getDeploymentSchema(project, fs);
  expect(schema).toMatchSnapshot();
});

test('getDeploymentSchema ', () => {
  const {
    project, vol, fs, table,
  } = makeProject();

  const schema = getDeploymentSchema(project, fs);

  expect(schema).toMatchSnapshot();
});
