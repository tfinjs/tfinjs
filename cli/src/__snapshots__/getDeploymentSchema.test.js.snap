// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDeploymentSchema  1`] = `
Object {
  "add": Array [],
  "graph": Object {
    "tijpetshop19qdr42": Array [],
    "tijpetshopmjubxt": Array [],
  },
  "remove": Array [],
  "resources": Object {
    "tijpetshop19qdr42": Object {
      "contentHash": "1ib4scr",
      "uri": "pet-shop/prod/static/aws/13371337/eu-north-1/customers/aws_dynamodb_table/customers",
    },
    "tijpetshopmjubxt": Object {
      "contentHash": "1jiyiui",
      "uri": "pet-shop/_/_/aws/13371337/eu-central-1/__tfinjs__backend__/aws_s3_bucket/terraform_state_prod",
    },
  },
  "update": Array [],
}
`;

exports[`getDeploymentSchema 1`] = `
Object {
  "add": Array [],
  "graph": Object {
    "tijpetshop19qdr42": Array [],
    "tijpetshopmjubxt": Array [],
  },
  "remove": Array [],
  "resources": Object {
    "tijpetshop19qdr42": Object {
      "contentHash": "1ib4scr",
      "uri": "pet-shop/prod/static/aws/13371337/eu-north-1/customers/aws_dynamodb_table/customers",
    },
    "tijpetshopmjubxt": Object {
      "contentHash": "1jiyiui",
      "uri": "pet-shop/_/_/aws/13371337/eu-central-1/__tfinjs__backend__/aws_s3_bucket/terraform_state_prod",
    },
  },
  "update": Array [],
}
`;
